dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java

```diff
@@ -0,0 +1,91 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class DeprecationWarningChannel implements Channel {
+    private static final Logger log = LoggerFactory.getLogger(DeprecationWarningChannel.class);
+    private static final Object SENTINEL = new Object();
+
+    private final Channel delegate;
+    private final TaggedMetricRegistry registry;
+    private final Cache<String, Object> loggingRateLimiter = Caffeine.newBuilder()
+            .expireAfterWrite(Duration.ofMinutes(1))
+            .build();
+
+    DeprecationWarningChannel(Channel delegate, TaggedMetricRegistry registry) {
+        this.delegate = delegate;
+        this.registry = registry;
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return DialogueFutures.addDirectCallback(
+                delegate.execute(endpoint, request),
+                DialogueFutures.onSuccess(response -> {
+                    if (response != null) {
+                        response.getFirstHeader("deprecation").ifPresent(deprecated -> {
+                            registry.meter(MetricName.builder()
+                                    .safeName("client.deprecations")
+                                    .putSafeTags("service-name", endpoint.serviceName())
+                                    .build())
+                                    .mark();
+
+                            if (tryAcquire(endpoint.serviceName())) {
+                                log.warn(
+                                        "Using a deprecated endpoint when connecting to service",
+                                        SafeArg.of("serviceName", endpoint.serviceName()),
+                                        SafeArg.of("endpointHttpMethod", endpoint.httpMethod()),
+                                        SafeArg.of("endpointName", endpoint.endpointName()),
+                                        SafeArg.of("endpointClientVersion", endpoint.version()),
+                                        SafeArg.of(
+                                                "service",
+                                                response.getFirstHeader("server").orElse("no server header provided")));
```

Comment:
"server" or "service"?

Reply:
the header is "server", but for logging purposes I think we want to call this parameter "service".